custom_ops.py

```diff
@@ -0,0 +1,78 @@
+import torch
+import numpy as np
+# aka torch.library
+from library import OpDef, traceable, device_types
+
+# =====================================================================
+# This is example user library code. It defines
+# my_sin and my_sin_cos operators.
+# =====================================================================
+
+# User provides their custom op schema and implementations
+class MySin(OpDef):
+    schema = "(Tensor x) -> Tensor"
+
+    # the black-box cpu kernel
+    @staticmethod
+    def impl_cpu(x):
+        return torch.from_numpy(np.sin(x.detach().cpu().numpy()))
+
+    # the black-box cuda kernel
+    @staticmethod
+    def impl_cuda(x):
+        return torch.from_numpy(np.sin(x.detach().cpu().numpy())).to(x.device)
+
+    # the abstract impl. Must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def abstract(x):
+        return torch.empty_like(x)
```

Comment:
This is not really a "new API" problem, but do I get fake tensors or meta tensors as inputs to abstract?

Reply:
(In particular, am I able to sniff the device on the tensors in order to vary striding layout on different devices?)

Reply:
You get FakeTensors if we're doing Fake propagation and meta tensors if we're doing operations with meta tensors. So yes.

I've been wondering if we need separate per-backend abstracts (e.g. abstract_cpu, abstract_cuda, ...), since there was an issue where a user wanted a way to add their own custom device and specify a custom abstract for it.

Reply:
They would need to call .build() again after doing the above

Reply:
we can get rid of build if the user doesn't change stuff